libraries/node-core-library/src/Async.ts

```diff
@@ -17,6 +17,21 @@ export interface IAsyncParallelismOptions {
   concurrency?: number;
 }
 
+/**
+ * Options for controlling the parallelism of asynchronous operations that have a property `weight?: number`.
+ *
+ * @remarks
+ * Used with {@link Async.forEachAsync}.
+ *
+ * @beta
+ */
+export interface IAsyncParallelismOptionsWithWeight extends IAsyncParallelismOptions {
+  /**
+   * If set, then will use `item.weight` instead of `1` as the parallelism consumed by a given iteration.
```

Comment:
This needs more explanation. I had to read the code carefully to figure out what `weight` actually represents. My guess was that somehow affected the priority of the task (like Unix `nice`), but that was wrong.  Also, the PR description and code comments never mention why someone would want this feature, i.e. what problem it solves.

Reply:
Suggested rewording:

```ts
  /**
   * Optionally used with the  {@link Async.mapAsync} and {@link Async.forEachAsync}
   * to limit amount of concurrency.
   * @remarks
   * By default, `concurrency` simply specifies the maximum number of promises that can be active
   * simultaneously.  However if `useWeight` is true, then behavior 
   * is different. See the {@link IAsyncParallelismOptions.useWeight} documentation for details.
   */
  concurrency?: number;

  /**
   * Set `useWeight` to true to configure the {@link IAsyncParallelismOptions.concurrency} 
   * setting to limit the total resource cost instead of limiting the total number of promises.
   * @remarks
   * By default, the `concurrency` specifies the maximum number of promises that can be
   * active simultaneously.  If promises have different resource costs, and those costs are known
   * in advance, then `useWeight` changes `concurrency` so that it limits the sum of the weights
   * of the entries passed to {@link Async.mapAsync} or {@link Async.forEachAsync}.
   * 
   * The {@link IAsyncEntry.weight} property is optional and defaults to 1.  (If all entries
   * use the default of 1, then the behavior is identical to `useWeight=false`.)
   * Weights can be fractional but not negative.
   */
  useWeight?: boolean;
```

```ts
  /**
   * If {@link Async.mapAsync} or {@link Async.forEachAsync} are used to parallelize tasks
   * that have different resource costs, and this property can be used to specify the resource
   * cost so that it can be limited by {@link IAsyncParallelismOptions.concurrency}.
   * @remarks
   * This property is ignored unless `useWeight` is true.  See the {@link 
   * IAsyncParallelismOptions.useWeight} documentation for details.
   */
  weight?: number
```

Reply:
Is `weight` the best word?  Maybe `concurrencyResourceCost` (or at least `concurrencyWeight`)?  It would be great of the naming gave some clue that `concurrency` and `weight` are part of the same feature.

Also we are asking to assign `weight` directly to somebody else's class/interface, so the more unique this name is, the less likely it is to accidentally conflict with an unrelated member.

Reply:
The name `weight` was chosen purely to match its existing presence on Rush's `Operation` class, but that could be changed to whatever name seems to make the most sense.

I'm also open to having an altogether separate entry point for tasks whose CPU resource needs are not uniformly 1.